server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -605,8 +613,56 @@ public void process(Version indexCreatedVersion, @Nullable MappingMetadata mappi
     }
 
     /* resolve the routing if needed */
-    public void resolveRouting(Metadata metadata) {
-        routing(metadata.resolveWriteIndexRouting(routing, index));
+    public void resolveRouting(
+        Metadata metadata,
+        IndexAbstraction abstraction,
+        Function<IndexMetadata, TimeSeriesIdGenerator> timeSeriesGeneratorLookup
+    ) {
+        // TODO clean this up once we tsid is its own field
+        String routingFromAliasOrRequest = metadata.resolveWriteIndexRouting(routing(), index);
+        boolean inTimeSeriesMode = abstraction == null || abstraction.getWriteIndex() == null
+            ? false
+            : abstraction.getWriteIndex().inTimeSeriesMode();
+        if (inTimeSeriesMode) {
+            if (routing() != null) {
+                throw new IllegalArgumentException(
+                    "routing cannot be set because the destination index [" + abstraction.getName() + "] is in time series mode"
+                );
+            }
+            if (routingFromAliasOrRequest != null) {
+                throw new IllegalArgumentException(
+                    "alias routing incompatible the destination index [" + abstraction.getName() + "] because it is in time series mode"
+                );
+            }
+            routing(routingFromTimeSeries(abstraction, timeSeriesGeneratorLookup));
```

Comment:
I'd like to put this into its own field and modify the routing behavior to pick it up. But I've not done that in this PR because I think it's fairly mechanical and change and I believe I'd have to disable a lot of BWC tests for it. I'd like to do that in a follow up. So instead I'm hiding the tsid in the _routing field, base 64 encoding it. Ewwwwww. But it works and we can zap it in a follow up.

Reply:
Maybe you already looked into it, but this might require some changes for `_split` to keep working with time series indices if we did that. `_split` looks at the value of the `_routing` field to know how to split shards in a way that still honors the rounting key (see `ShardSplittingQuery`).

Reply:
OOOH! I hadn't thought of it. You've given me a wonderful thing to experiment with. I imagine I break this now. I'll look!

Reply:
Shard splitting query is indeed busted here. We'll need revive it, but I'd like to wait for a follow up for that.